### DISCUSSION TITLE
Android demo app: Multilingual support?

### DISCUSSION THREAD

**Discussion by darlingtonplate** [ggml-org/whisper.cpp#549] [Q&A] [answered]
Hello everyone. Today, I successfully was able to launch the android  demo app on my physical device, however I noticed one thing: it doesn't seem to support any languages other than transcribing in English.

Either that, or I'm blind as I am particularly new to programming for Android. I'd also like to mention that the `README.md` file within [examples/whisper.andriod](https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.android) fails to mention the requirement of downloading and unzipping the contents of `examples/whisper` from the [ggml repository](https://github.com/ggerganov/ggml/tree/master/examples/whisper). Unless if this was an issue with my implementation.

**Comment by fuhadabdulla** [comment]
Hey , Can you please tell me how to build the android app . I have no experience in android development . If you can please share the apk

**Reply by darlingtonplate** [reply]
Hey there, it is quite easy.

First of all, you should have all the essentials prepped (Android Studio and Git installed) which then afterwards you can easily fork the Android example and build it for the first time.

Then, you should copy/paste all the contents within `examples/whisper` from [here](https://github.com/ggerganov/ggml/tree/master/examples/whisper). Attempt building it this time, works usually like a charm.

**Reply by jaimebw** [reply]
Do you need to put the content of the `examples/whisper` inside any specific directory of the app? I'm pretty new to Android programming too. PS: how much time does it take to run the transcription on the example file on your system?

**Reply by darlingtonplate** [reply]
I apologize for not adding enough information to my last reply, yes 

You'll need to extract the contents of `examples/whisper` into the directory `app/src/main/jni`
Took me an hour just spinning around myself to find this 😅 If it doesn't work, please share your folder hierarchy and I'll try to figure something out :)

Oh, and do not forget to create an `assets/models` & `assets/samples` folder as they are not created (simply follow the instructions in the readme.md from there on)

**Reply by darlingtonplate** [reply]
On my Galaxy Z Flip 4 it takes around 7~ seconds to 30~ seconds depending on the length of the audio file. For a short, 5 second audio recording - it took about 7 seconds so make with that as you will.

I did figure out however that running the base `main` version of whisper.cpp on an ubuntu sever provided better results (it was tested on ARM which is pretty much similar to what's on a phone)

**Reply by jaimebw** [reply]
Thanks for the help and the benchmark times. I've been able to build the app correctly yet the emulator seems incapable of running the model( with the base and the tiny models). I will try to run it on an old Samsung S9 and see how it goes. 
Btw, I think that the issue with the build and having to add the ```examples/whisper``` library is due to this: 

https://github.com/ggerganov/whisper.cpp/blob/72af0f56975bfea20ff340777b3c940b006bf42a/examples/whisper.android/app/src/main/jni/whisper/jni.c#L8

Basically, idk why but Android studio changes the path of the file to:
```c
#include "examples/whisper.android/app/src/main/jni/whisper.h"
```
And provokes all the issues with the build. 
@tinoue do you know why this might be happening?

**Reply by tinoue** [reply]
@jaimebw I have no include path problem. The include path is specified here:
https://github.com/ggerganov/whisper.cpp/blob/72af0f56975bfea20ff340777b3c940b006bf42a/examples/whisper.android/app/src/main/jni/whisper/Whisper.mk#L14-L18

Make sure you open `whisper.cpp/examples/whisper.android` from Android Studio.

**Reply by darlingtonplate** [reply]
> Btw, I think that the issue with the build and having to add the `examples/whisper` library is due to this:

Yeah, I think so too. Previous builds refused to build without having the library.
I haven't tested it on virtualized hardware, but I'd recommend you to test it on physical hardware either way, as it it more representative of actual performance

**Reply by jaimebw** [reply]
FYI, just tested in a Samsung Galaxy S8 and it took 1 minute and 25 seconds to transcribe the sample audio using the tiny model. Thanks again for all the help 😊😊

**Reply by jaimebw** [reply]
Try to add some [Toast](https://developer.android.com/reference/android/widget/Toast), or use the [log](https://developer.android.com/reference/android/util/Log) library in the example app to see what's going on.

**Comment by tinoue** [comment, marked as answer]
@darlingtonplate For now you should hard-code 2-characters language parameter `params.language` at:
https://github.com/ggerganov/whisper.cpp/blob/master/examples/whisper.android/app/src/main/jni/whisper/jni.c#L182

Available languages: https://github.com/ggerganov/whisper.cpp/blob/master/whisper.cpp#L119

**Reply by darlingtonplate** [reply]
Spectacular! Thanks, I felt like I wasn't searching in the correct place.

**Reply by helpforsoftware** [reply]
is Turkish? "TR" supported? links are dead. 

i tried to run with turkish  language  but it is not working. (changed jni.c file )

**Reply by ozanmuyes** [reply]
[Fixed link](https://github.com/ggml-org/whisper.cpp/blob/999a7e0cbf8484dc2cea1e9f855d6b39f34f7ae9/src/whisper.cpp#L280) for "Available languages" (please note the commit hash, the list might be changed when you visit)

Turkish seems to be one of the available languages and considering that ["Models are multilingual unless the model name includes `.en`."](https://github.com/ggml-org/whisper.cpp/blob/999a7e0cbf8484dc2cea1e9f855d6b39f34f7ae9/models/README.md?plain=1#L63) any model without `".en"` in its name theoretically should work.

Although I'm not trying to run whisper.cpp on Android, this is the only thing I can find about whisper.cpp and Turkish. I just wanted to share my 2 cents.

**Comment by sudipa23** [comment]
I've tried to build the apk from the android demo of whisper.cpp. but I got the following error : [CXX1405] error when building with ndkBuild using D:\whisper_cpp\whisper.android\app\src\main\jni\whisper\Android.mk: C++ build system [configure] failed while executing:

can anyone please help me on that

**Reply by rjpikus** [reply]
I wonder if there is a c++ config file, or c++ libs needed, I do not know, dealing my errors and will keep this in mind. Forging ahead.

**Reply by sudipa23** [reply]
finally worked in my android phone. Need to set up my NDK location.as the code uses C++ and in android we need NDK toolset to deal with  the code written in C++ .

**Reply by rjpikus** [reply]
Explain how you got working? 
Add the files from ggml to jni?
or no add files from ggml?

**Reply by sudipa23** [reply]
as I mentioned earlier that I got an issue in my NDK location and that also caused because I had white space in my SDK location path, so the NDK location file path can not accessing that even after installing NDK(step by step ) after fixing this issue ,the build is successful. But even after build is successful the app is not recognizing anything then I figure out that the model (ggml-base.bin) file need to download from its original site which is https://huggingface.co/ggerganov/whisper.cpp/blob/main/ggml-base.bin. the models that are uploaded in this repo is not the actual models it's for demo.

**Comment by tumuyan** [comment]
I have made a ci script to build Android demo with model & language choose.

https://github.com/tumuyan/whisper.cpp/blob/master/.github/workflows/build-android.yml

![image](https://user-images.githubusercontent.com/3126801/232473580-e8255bd8-3f9f-4a22-bfe3-b2b6bb24038a.png)

**Reply by helpforsoftware** [reply]
is it still working?

**Reply by tumuyan** [reply]
no, something different
